src/MatrixFields/field_matrix_iterative_solver.jl

```diff
@@ -0,0 +1,465 @@
+    # Turn all boolean fields into type parameters to ensure type-stability.
+    params = (correlated_solves, print_norm, print_radius, typeof(P_alg), K)
+    return StationaryIterativeSolve{params...}(P_alg, n_iters, eigsolve_kwargs′)
+end
+
+Base.getproperty(
+    alg::StationaryIterativeSolve{correlated_solves, print_norm, print_radius},
+    name::Symbol,
+) where {correlated_solves, print_norm, print_radius} =
+    if name == :correlated_solves
+        correlated_solves
+    elseif name == :print_norm
+        print_norm
+    elseif name == :print_radius
+        print_radius
+    else
+        getfield(alg, name)
+    end # Extract the boolean type parameters as if they were regular fields.
+
+function field_matrix_solver_cache(alg::StationaryIterativeSolve, A, b)
+    P_cache = preconditioner_cache(alg.P_alg, A, b)
+    # Note: We cannot use similar_to_x here because it doesn't work for some
+    # particularly complicated unit tests. For now, we will assume that x is
+    # similar to b, rather than just keys(x) == keys(b).
+    previous_x_cache =
+        alg.correlated_solves ? (; previous_x = zero.(similar(b))) : (;)
+    return (; P_cache, previous_x_cache...)
+end
+
+check_field_matrix_solver(alg::StationaryIterativeSolve, cache, A, b) =
+    check_preconditioner(alg.P_alg, cache.P_cache, A, b)
+
+function run_field_matrix_solver!(alg::StationaryIterativeSolve, cache, x, A, b)
+    P = concrete_preconditioner(alg.P_alg, cache.P_cache, A)
+    if alg.print_radius
+        e₀ = concrete_field_vector(b) # Initialize e to any nonzero vector.
+        eigenvalues, _, info = eigsolve(e₀, 1; alg.eigsolve_kwargs...) do e
+            e_view = field_vector_view(e, keys(b).name_tree)
+            lazy_A_e = lazy_mul(A, e_view)
+            lazy_invP_A_e =
+                apply_preconditioner(alg.P_alg, cache.P_cache, P, lazy_A_e)
+            concrete_field_vector(@. e_view - lazy_invP_A_e)
+        end
+        if info.converged == 0
+            (; tol, maxiter) = alg.eigsolve_kwargs
+            @warn "Unable to approximate ρ(I - inv(P) * A) to within a \
+                   tolerance of $(100 * tol) % in $maxiter or fewer iterations"
+        else
+            spectral_radius = abs(eigenvalues[1])
+            if spectral_radius < 1
+                @info "ρ(I - inv(P) * A) ≈ $spectral_radius"
+            else
+                @warn "StationaryIterativeSolve may not converge because \
+                       ρ(I - inv(P) * A) ≈ $spectral_radius is not less than 1"
+            end
+        end
+    end
+    if alg.correlated_solves
+        @. x = cache.previous_x
+    else
+        @. x = zero(x)
+    end
+    for iter in 1:(alg.n_iters)
+        lazy_Δb = lazy_sub(b, lazy_mul(A, x))
+        lazy_Δx = apply_preconditioner(alg.P_alg, cache.P_cache, P, lazy_Δb)
+        if alg.print_norm
+            norm_Δx = norm(concrete_field_vector(Base.materialize(lazy_Δx)))
+            @info "||x[$(iter - 1)] - x'||₂ ≈ $norm_Δx"
+        end
```

Comment:
Technically you're logging, not printing. A better way to control this is to do something like
```suggestion
        @debug "||x[$(iter - 1)] - x'||₂ ≈ $(norm(concrete_field_vector(Base.materialize(lazy_Δx))))"
```
Then you can switch it on by setting the environment variable `JULIA_DEBUG=MatrixFields`.

Reply:
With this strategy, is there any way to control the norm debugging and the radius debugging individually? Computing the error norm is very cheap, so we'll probably end up printing that somewhat frequently. The spectral radius takes a very long time to compute, though, so we'll only need to check it occasionally.

Reply:
JULIA_DEBUG accepts filenames, modules or groups:
```
➜  ~ JULIA_DEBUG=mygroup julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.9.4 (2023-11-14)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> @debug "hello"

julia> @debug "hello" _group=:mygroup
┌ Debug: hello
└ @ Main REPL[4]:1
```

So you could use a custom group for the different things you want to log?